docs/naps/0-nap-process.md

```diff
@@ -0,0 +1,272 @@
+---
+Author: "Juan Nunez-Iglesias <mailto:jni@fastmail.com>"
+Created: '2022-03-23'
+Status: Active
+Type: Process
+---
+
+(nap0)=
+
+# NAP 0 — Purpose and Process
+
+## What is a NAP?
+
+NAP stands for Napari Advancement Proposal. A NAP is a design document
+providing information to the community, or describing a new feature for
+napari or its processes or environment. The NAP should provide a rationale
+for the proposed change as well as a concise technical specification, if
+applicable.
+
+We intend NAPs to be the primary mechanisms for proposing major new
+features, for collecting community input on an issue, and for documenting
```

Comment:
I think one of the main motivations behind NAPs is to prevent people from putting too much work into something before a PR or set of PRs is ready for review - i.e. it's partly for the author/contributor to avoid wasting time/resources on something the community doesn't want. If the PR is ready and concise enough to review, then retroactively going through the NAP process seems less useful, unless there are strong objections to the PR.

I guess one of the other main motivations is to help document decisions, processes, and designs. So this might be one case in which retroactively writing a NAP for an existing PR might make sense - **more so** if the PR was ultimately closed because documentation of a rejected proposal is a really clear/quick way of deciding what we shouldn't work on in the future.

Reply:
There's several situations that might warrant a NAP, and they are all different, so it would indeed be good to enumerate them.

1. I think this is the most common: someone proposes a new feature, or a change in behaviour, or a change in process. This is done informally, either through an issue, or a new thread in Zulip. Often the response is unanimous ("Great idea! Can you open a PR?"), and then things can just go ahead. Some times, though, there is debate about whether the approach is a good one, or whether there's a different approach that might be better, or whether it's even in-scope for the project. In those situations, *even if consensus is reached*, a NAP would be warranted, in order to record the arguments that were considered for or against a particular approach. This is extremely important to avoid revisiting the same design questions over and over.
2. Same as above, but *consensus is **not** reached* in the initial discussion. At that point, without NAPs, usually what happens is that momentum is lost, and the idea remains in perpetual limbo. NAPs are a good mechanism to force a decision to be taken one way or another. The NAP should contain the arguments for or against, with links to the discussion, and then the core devs, or in extreme circumstances the steering council, can go ahead and make a decision, and work can proceed in one direction or another.
3. Finally, in many cases, design decisions are taken without much thought, or failing to consider particular use cases (for example from communities that at the time were not well-represented in the napari development community — e.g. our transforms API being so geared towards subsets of affine transforms). In those situations, discussion may arise as to why a particular design is one way. If we discuss and decide to keep behaviour unchanged, an Informational NAP could be warranted. If we decide to change things, then it reduces to case 1 or 2 and we go from there.

I do think 3 comes up pretty often, and I think that falls into what you would call "retroactive", @andy-sweet, and it can be a really good way to build up documentation around things that have been tried/not tried/dismissed.

I'll also point out that NAPs can be superseded, e.g. Python had several rejected proposals for Protocols/Structural Typing before one was accepted.

@kevinyamauchi as to what constitutes a "major" feature, I suggest that anything that expands the viewer API is one. `_experimental_` features wouldn't be. There's fuzzy ground in the middle and to be honest I think we need to use judgement there rather than have strong definitions.

Reply:
I've pushed a new section about when NAPs are warranted, feel free to edit further!

Also, @andy-sweet and @kevinyamauchi, could you add yourselves as Authors in the top section? I didn't do it because I didn't know which emails to use (if any), but you have now made important changes to this document! I'd be tempted to backport them to skimage and numpy if it wasn't going to be a massive uphill battle :joy:

Reply:
Done. Thanks for the inclusion!

Reply:
> Same as above, but consensus is not reached in the initial discussion. At that point, without NAPs, usually what happens is that momentum is lost, and the idea remains in perpetual limbo. NAPs are a good mechanism to force a decision to be taken one way or another. 

@jni could you expand on this a little? I don't immediately see how lack of consensus around an enhancement proposal is resolved with the NAP process, is there some time limit after which a NAP must be accepted, rejected or deferred?

Reply:
There's no strict time limit, but there is a mechanism to push it through or reject it:

> If there *are* substantive objections, then the NAP remains in
>`Draft` state, discussion continues as normal, and it can be
> proposed for acceptance again later once the objections are resolved.
> 
> In unusual cases, when no consensus can be reached between core developers,
> the [napari Steering Council] may be asked to decide whether a
> controversial NAP is `Accepted`.

Reply:
> Also, @andy-sweet and @kevinyamauchi, could you add yourselves as Authors in the top section?

Done. Thanks!

Reply:
> I've pushed a new section about when NAPs are warranted, feel free to edit further!

Thank you, I think this clarifies things a lot!